spring-integration-core/src/main/java/org/springframework/integration/gateway/MessagingGatewaySupport.java

```diff
@@ -533,6 +546,135 @@ private Object doSendAndReceive(Object object, boolean shouldConvert) {
+	}
+
+	@SuppressWarnings("unchecked")
+	private Mono<Message<?>> doSendAndReceiveMessageReactive(MessageChannel requestChannel, Object object,
+			boolean error) {
+
+		AtomicReference<Message<?>> requestMessageReference = new AtomicReference<>();
+
+		return Mono
+				.<Message<?>>defer(() -> {
+					Message<?> requestMessage;
+					try {
+						requestMessage = object instanceof Message<?>
+								? (Message<?>) object
+								: this.requestMapper.toMessage(object);
+
+						requestMessage = this.historyWritingPostProcessor.postProcessMessage(requestMessage);
+
+					}
+					catch (Exception e) {
+						throw new MessageMappingException("Cannot map to message: " + object, e);
+					}
+
+					Object originalReplyChannelHeader = requestMessage.getHeaders().getReplyChannel();
+					Object originalErrorChannelHeader = requestMessage.getHeaders().getErrorChannel();
+
+					FutureReplyChannel replyChannel = new FutureReplyChannel();
+
+					requestMessage = MutableMessageBuilder.fromMessage(requestMessage)
+							.setReplyChannel(replyChannel)
+							.setHeader(this.messagingTemplate.getSendTimeoutHeader(), null)
+							.setHeader(this.messagingTemplate.getReceiveTimeoutHeader(), null)
+							.setErrorChannel(replyChannel)
+							.build();
+
+					requestMessageReference.set(requestMessage);
+
+					if (requestChannel instanceof ReactiveStreamsSubscribableChannel) {
+						((ReactiveStreamsSubscribableChannel) requestChannel)
+								.subscribeTo(Mono.just(requestMessage));
+					}
+					else {
+						long sendTimeout = sendTimeout(requestMessage);
+
+						boolean sent =
+								sendTimeout >= 0
+										? requestChannel.send(requestMessage, sendTimeout)
+										: requestChannel.send(requestMessage);
+
+						if (!sent) {
+							throw new MessageDeliveryException(requestMessage,
+									"Failed to send message to channel '" + requestChannel +
+											"' within timeout: " + sendTimeout);
+						}
+					}
+
+					return Mono.fromFuture(replyChannel.messageFuture)
+							.map(replyMessage ->
+									MessageBuilder.fromMessage(replyMessage)
+											.setHeader(MessageHeaders.REPLY_CHANNEL, originalReplyChannelHeader)
+											.setHeader(MessageHeaders.ERROR_CHANNEL, originalErrorChannelHeader)
+											.build());
+				})
+				.doOnSubscribe(s -> {
+					if (!error && this.countsEnabled) {
+						this.messageCount.incrementAndGet();
+					}
+				})
+				.onErrorResume(t -> handleSendError(requestMessageReference.get(), t));
+	}
+
+	private Mono<Message<?>> handleSendError(Message<?> requestMessage, Throwable exception) {
+		if (logger.isDebugEnabled()) {
+			logger.debug("failure occurred in gateway sendAndReceiveReactive: " + exception.getMessage());
+		}
+		MessageChannel errorChannel = getErrorChannel();
+		if (errorChannel != null) {
+			ErrorMessage errorMessage = buildErrorMessage(requestMessage, exception);
+			try {
+				return doSendAndReceiveMessageReactive(errorChannel, errorMessage, true);
```

Comment:
Does it resume the response from error channel then ? If so the message will be different, is it expected ?